File: typescript/grimlock/versions/v1/recovery-key.ts
/**
 * V1 Recovery Key Operations for Grimlock crypto module
 * 
 * Implements recovery key generation and encoding.
 */

import type { RecoveryKey } from '../../types/common';
import { CRYPTO_CONSTANTS_V1 } from './constants';
import { base64Encode, base64Decode } from '../../utils/encoding';

/**
 * Generate a cryptographically secure random 32-byte array
 */
function getRandomBytes(length: number): Uint8Array {
  // Use Web Crypto API (available in browser and Next.js)
  if (typeof globalThis !== 'undefined' && (globalThis.crypto?.getRandomValues instanceof Function)) {
    const array = new Uint8Array(length);
    globalThis.crypto.getRandomValues(array);
    return array;
  } else {
    throw new Error('Web Crypto API is required but not available');
  }
}

/**
 * Generate a recovery key
 * 
 * Generates a 256-bit (32-byte) cryptographically secure random key
 * and encodes it in multiple formats for user convenience.
 * 
 * @returns Recovery key with raw bytes, Base64, and optional mnemonic
 */
export function generateRecoveryKey(): RecoveryKey {
  // Generate 32 random bytes (256 bits)
  const raw = getRandomBytes(CRYPTO_CONSTANTS_V1.recoveryKeySize);

  // Encode as Base64 (44 characters)
  const base64 = base64Encode(raw);

  // Optional: Generate BIP39 mnemonic (requires bip39 package)
  // For now, we'll leave it undefined. Users can install bip39
  // and generate mnemonic separately if needed.
  const mnemonic = undefined; // TODO: Add BIP39 support if needed

  return {
    raw,
    base64,
    mnemonic,
  };
}

/**
 * Parse recovery key from Base64 string
 * 
 * @param base64Key - Base64 encoded recovery key
 * @returns Recovery key with raw bytes
 */
export function parseRecoveryKeyFromBase64(base64Key: string): RecoveryKey {
  const raw = base64Decode(base64Key);

  if (raw.length !== CRYPTO_CONSTANTS_V1.recoveryKeySize) {
    throw new Error(
      `Invalid recovery key size: expected ${CRYPTO_CONSTANTS_V1.recoveryKeySize} bytes, got ${raw.length}`
    );
  }

  return {
    raw,
    base64: base64Key,
    mnemonic: undefined,
  };
}

/**
 * Parse recovery key from BIP39 mnemonic
 * 
 * This requires the bip39 package to be installed.
 * 
 * @param mnemonic - BIP39 mnemonic phrase (24 words)
 * @returns Recovery key with raw bytes
 */
export function parseRecoveryKeyFromMnemonic(mnemonic: string): RecoveryKey {
  // This is a placeholder - requires bip39 package
  // 
  // Example implementation:
  // const bip39 = require('bip39');
  // const entropy = bip39.mnemonicToEntropy(mnemonic);
  // const raw = hexDecode(entropy);
  //
  // if (raw.length !== CRYPTO_CONSTANTS_V1.recoveryKeySize) {
  //   throw new Error('Invalid mnemonic entropy length');
  // }
  //
  // return {
  //   raw,
  //   base64: base64Encode(raw),
  //   mnemonic,
  // };
  
  throw new Error(
    'BIP39 mnemonic parsing requires bip39 package. ' +
    'Please install it: npm install bip39'
  );
}

File: python/grimlock/versions/v1/recovery_key.py
"""Recovery key generation for Grimlock crypto module."""

import os
from mnemonic import Mnemonic

from ...types.common import RecoveryKey
from ...utils.encoding import encode_base64
from .constants import CRYPTO_CONSTANTS_V1
from .encryption import decrypt_private_key, encrypt_private_key
from .key_derivation import derive_recovery_key


def generate_recovery_key() -> RecoveryKey:
    """Generate a new cryptographically secure recovery key.

    Returns:
        RecoveryKey: Recovery key with raw bytes, base64 encoding, and optional mnemonic

    Raises:
        ValueError: If key generation fails
    """
    # Generate random 256-bit key
    key = os.urandom(CRYPTO_CONSTANTS_V1.recovery_key_size)

    # Generate BIP39 mnemonic (optional)
    try:
        mnemo = Mnemonic("english")
        mnemonic = mnemo.to_mnemonic(key)
    except Exception:
        # If mnemonic generation fails, continue without it
        mnemonic = None

    return RecoveryKey(
        raw=key,
        base64=encode_base64(key),
        mnemonic=mnemonic,
    )


def encrypt_private_key_with_recovery_key(
    private_key: bytes, recovery_key: bytes, aad: bytes
):
    """Encrypt a private key using a recovery key.

    Args:
        private_key: Private key to encrypt (32 bytes)
        recovery_key: Recovery key bytes (32 bytes)
        aad: Additional authenticated data

    Returns:
        EncryptedPrivateKeyV1: Encrypted private key

    Raises:
        ValueError: If recovery key size is invalid
    """
    if len(recovery_key) != CRYPTO_CONSTANTS_V1.recovery_key_size:
        raise ValueError(
            f"invalid recovery key size: expected {CRYPTO_CONSTANTS_V1.recovery_key_size}, "
            f"got {len(recovery_key)}"
        )

    # Derive encryption key from recovery key
    encryption_key = derive_recovery_key(recovery_key)

    try:
        # Encrypt private key
        return encrypt_private_key(private_key, encryption_key, aad)
    finally:
        # Best-effort secure erase
        if isinstance(encryption_key, bytearray):
            encryption_key[:] = b"\x00" * len(encryption_key)


def decrypt_private_key_with_recovery_key(
    encrypted, recovery_key: bytes, aad: bytes
) -> bytes:
    """Decrypt a private key using a recovery key.

    Args:
        encrypted: Encrypted private key
        recovery_key: Recovery key bytes (32 bytes)
        aad: Additional authenticated data

    Returns:
        bytes: Decrypted private key

    Raises:
        ValueError: If recovery key size is invalid
    """
    if len(recovery_key) != CRYPTO_CONSTANTS_V1.recovery_key_size:
        raise ValueError(
            f"invalid recovery key size: expected {CRYPTO_CONSTANTS_V1.recovery_key_size}, "
            f"got {len(recovery_key)}"
        )

    # Derive encryption key from recovery key
    encryption_key = derive_recovery_key(recovery_key)

    try:
        # Decrypt private key
        return decrypt_private_key(encrypted, encryption_key, aad)
    finally:
        # Best-effort secure erase
        if isinstance(encryption_key, bytearray):
            encryption_key[:] = b"\x00" * len(encryption_key)


def validate_recovery_key(recovery_key: bytes) -> None:
    """Validate that a recovery key is well-formed.

    Args:
        recovery_key: Recovery key bytes to validate

    Raises:
        ValueError: If recovery key is invalid
    """
    if len(recovery_key) != CRYPTO_CONSTANTS_V1.recovery_key_size:
        raise ValueError(
            f"invalid recovery key size: expected {CRYPTO_CONSTANTS_V1.recovery_key_size}, "
            f"got {len(recovery_key)}"
        )

    # Check that recovery key is not all zeros
    all_zero = all(b == 0 for b in recovery_key)
    if all_zero:
        raise ValueError("recovery key cannot be all zeros")
